rustls/src/crypto/hpke.rs

```diff
@@ -0,0 +1,131 @@
+
+    /// Does the provider support the given mode?
+    fn supports_mode(&self, mode: Mode) -> bool;
+
+    /// Does the provider support the given [HpkeKem]?
+    fn supports_kem(&self, kem: HpkeKem) -> bool;
+
+    /// Does the provider support the given [HpkeKdf]?
+    fn supports_kdf(&self, kdf: HpkeKdf) -> bool;
+
+    /// Does the provider support the given [HpkeAead]?
+    fn supports_aead(&self, aead: HpkeAead) -> bool;
+}
+
+/// Available HPKE authentication modes.
+///
+/// The base case for HPKE is unauthenticated encryption to a public key. In addition to this
+/// mode there are three modes to provide a form of authentication.
+///
+/// See [RFC 9180 Section 5](https://www.rfc-editor.org/rfc/rfc9180.html#name-hybrid-public-key-encryptio)
+/// for more details.
+#[derive(PartialEq, Copy, Clone, Debug)]
+#[repr(u8)]
+pub enum Mode {
+    /// Encryption to a public key, no additional authentication.
+    Base = 0x00,
+
+    /// Encryption to a public key, authenticated with a pre-shared secret key.
+    Psk = 0x01,
+
+    /// Encryption to a public key, authenticated with the sender's private key.
+    Auth = 0x02,
+
+    /// Encryption to a public key, authenticated with the sender's private key and a
+    /// pre-shared secret key.
+    AuthPsk = 0x03,
+}
+
+/// An HPKE instance that can produce sender and receiver contexts.
+pub trait Hpke: Debug + Send + Sync {
+    /// Create a sender context to encrypt messages to a receiver.
+    ///
+    /// This will return an encapsulated secret and a sender context. The sender context
+    /// can be used to seal messages to the receiver.
+    fn setup_sender(
+        &self,
+        pk_r: &HpkePublicKey,
+        info: &[u8],
+        psk: Option<&HpkePreSharedKey>,
+        psk_id: Option<&[u8]>,
+        sk_s: Option<&HpkePrivateKey>,
+    ) -> Result<(EncapsulatedSecret, Box<dyn HpkeSender + '_>), Error>;
+
+    /// Create a receiver context to decrypt messages from a sender.
+    fn setup_receiver(
+        &self,
+        enc: &EncapsulatedSecret,
+        sk_r: &HpkePrivateKey,
+        info: &[u8],
+        psk: Option<&HpkePreSharedKey>,
+        psk_id: Option<&[u8]>,
+        pk_s: Option<&HpkePublicKey>,
+    ) -> Result<Box<dyn HpkeReceiver + '_>, Error>;
+}
+
+/// An HPKE Sender context.
+///
+/// This is a stateful object that can be used to seal messages for receipt by
+/// a receiver.
+pub trait HpkeSender: Debug + Send + Sync {
+    /// Seal the provided `plaintext` with additional data `aad`, returning
+    /// ciphertext.
+    fn seal(&mut self, aad: &[u8], plaintext: &[u8]) -> Result<Vec<u8>, Error>;
+}
+
+/// An HPKE Receiver context.
+///
+/// This is a stateful object that can be used to open sealed messages sealed
+/// by a sender.
+pub trait HpkeReceiver: Debug + Send + Sync {
+    /// Open the provided `ciphertext` with additional data `aad`, returning plaintext.
+    fn open(&mut self, aad: &[u8], ciphertext: &[u8]) -> Result<Vec<u8>, Error>;
```

Comment:
This introduces the usual issue with AEADs: "premature release of unauthenticated plaintext" (ie, the decryption cannot tell you whether it was tampered until the end, but by the time you have already received some of the corrupted plaintext). A single-shot API would avoid that.

Reply:
The latest push switched to the single-shot API.